Zipped NodeJS modules as actions

psuter opened the pull request:
This PR introduces support for Node actions written as zipped modules.

Builds on #1348 which should be merged first.

psuter commented:
PG2 381 is blue-ish

bjustin-ibm commented:
It might be nice to supply a [Yeoman generator](http://yeoman.io/learning/index.html) to help generate OpenWhisk actions.

bjustin-ibm commented:
Thinking longer term, it might be advisable to support .tgz files produced with the `npm pack` command. `npm pack` is smart enough to include "bundled" dependencies and exclude files that are not needed for production use (automated tests, build files, README, etc.).

This would require that `npm install` is run on the .tgz each time, however, if all the required dependencies are bundled, this command is essentially equivalent to just unpacking the .tgz. However, if all the required dependencies are not bundled, the installation will still work (albeit slower due to the overhead of fetching dependencies). At this point, using `bundledDependencies` is an optimization we can strongly encourage to the user through documentation, the aforementioned Yeoman generator, and even log messages when running actions that did not include all the required deps.

bjustin-ibm commented on core/nodejsActionBase/runner.js:
```diff
@@ -108,6 +127,55 @@ function NodeActionRunner(whisk) {
+        return exec(mkTempCmd).then(function (tmpDir1) {
+            return new Promise(
+                function (resolve, reject) {
+                    var zipFile = tmpDir1 + "/action.zip";
```
Using the `path` module to join paths is ever-so-slightly nicer:
```javascript
var zipFile = path.join(tmpDir1, 'action.zip')
```

bjustin-ibm commented on core/nodejsActionBase/runner.js:
```diff
@@ -47,26 +50,42 @@ function NodeActionRunner(whisk) {
+                thisRunner.userScriptMain = eval(message.main);
+                assertMainIsFunction()
+                // See comment above about 'true'; it has no specific meaning.
+                return Promise.resolve(true);
```
😒

psuter commented on core/nodejsActionBase/runner.js:
```diff
@@ -47,26 +50,42 @@ function NodeActionRunner(whisk) {
+                // See comment above about 'true'; it has no specific meaning.
+                return Promise.resolve(true);
+            } catch (e) {
+                return Promise.reject(e);
```
I did not know that! :smiley: Thanks for the education.

psuter commented on core/nodejsActionBase/runner.js:
```diff
@@ -108,6 +127,55 @@ function NodeActionRunner(whisk) {
+    var base64Regex = /^([0-9a-zA-Z+/])*=*$/;
+    function isBase64(str) {
+        return base64Regex.test(str);
+    }
```
Sounds good.

psuter commented on core/nodejsActionBase/runner.js:
```diff
@@ -21,6 +21,9 @@
 var util = require('util');
+var child_process = require('child_process');
+var fs = require('fs');
+var path = require('path');
```
Out of curiosity, why is that better? Is it a loading time issue?

bjustin-ibm replied:
Exactly. `require` immediately goes out to disk reads the file and sticks it in memory. Lazy loading modules saves both IO time and memory usage in paths where you don't call the `require`.

psuter replied:
It's a delicate balance then.. Assuming a warm js container, `require`ing at the top saves some latency when initializing a zip-based action. Assuming a cold container, it wastes some time for non-zip actions?

bjustin-ibm replied:
I believe what you are saying is correct. My gut feeling is that in all cases, we should delay this loading until we know we need it. If that initialization happens to get re-used by a warm container, then that is even better.

psuter commented on tests/src/actionContainers/NodeJsActionContainerTests.scala:
```diff
@@ -434,4 +436,60 @@ class NodeJsActionContainerTests extends BasicActionRunnerTests with WskActorSys
+                (o + e).toLowerCase should include("error")
+                (o + e).toLowerCase should include("archive")
+        })
+    }
```
Yes

psuter commented:
The way I see it, supporting `.tgz` could be part of this PR, but supporting `npm install` seems beyond its scope.

bjustin-ibm commented:
> The way I see it, supporting .tgz could be part of this PR, but supporting npm install seems beyond its scope.

I agree, and I even think that supporting `.tgz` without `npm install` may be more confusing than valuable. In this way, we may be able to say `.zip` is for completely self-contained archives, and `.tgz` is for using with `npm install`.

psuter commented:
Addressed the `binary` marker (#1391 needs to be merged first).

Improved the error messages for the various failure modes of uploading a broken/incomplete zip file.

Added tests for the above.

psuter commented:
PG2 415

rabbah commented:
Merged as https://github.com/openwhisk/openwhisk/compare/cb02a8850a4e...b36fd6faafbe.
